### PR TITLE
GH3714: Use Basic.Reference.Assemblies.* to ensure all reference assemblies are loaded

### LINKED ISSUE #3714: Use Basic.Reference.Assemblies.* to ensure all standard reference assemblies are available for roslyn
* .NET 6.0 - Basic.Reference.Assemblies.Net60
* .NET 5.0 - Basic.Reference.Assemblies.Net50
* .NET Core 3.1 - Basic.Reference.Assemblies.NetCoreApp31

### PR DESCRIPTION
**Pull request by devlead**
* .NET 6.0 - Basic.Reference.Assemblies.Net60
* .NET 5.0 - Basic.Reference.Assemblies.Net50
* .NET Core 3.1 - Basic.Reference.Assemblies.NetCoreApp31
* fixes #3714

### CODE REVIEW

File: src/Cake/Cake.csproj [hunk 1]
```diff
@@ -26,9 +26,12 @@
...
   <!-- Global packages -->
   <ItemGroup>
-    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Scripting" Version="4.0.0-6.final" />
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Scripting" Version="4.0.1" />
     <PackageReference Include="System.Collections.Immutable" Version="6.0.0" />
     <PackageReference Include="System.Reflection.Metadata" Version="6.0.0" />
     <PackageReference Include="Autofac" Version="6.3.0" />
+	<PackageReference Include="Basic.Reference.Assemblies.Net60" Version="1.2.4" Condition="'$(TargetFramework)' == 'net6.0'" />
+	<PackageReference Include="Basic.Reference.Assemblies.Net50" Version="1.2.4" Condition="'$(TargetFramework)' == 'net5.0'" />
+	<PackageReference Include="Basic.Reference.Assemblies.NetCoreApp31" Version="1.2.4" Condition="'$(TargetFramework)' == 'netcoreapp3.1'" />
```
Comment (gitfool) [inline src/Cake/Cake.csproj:35]:
```suggestion
    <PackageReference Include="Basic.Reference.Assemblies.Net60" Version="1.2.4" Condition="'$(TargetFramework)' == 'net6.0'" />
    <PackageReference Include="Basic.Reference.Assemblies.Net50" Version="1.2.4" Condition="'$(TargetFramework)' == 'net5.0'" />
    <PackageReference Include="Basic.Reference.Assemblies.NetCoreApp31" Version="1.2.4" Condition="'$(TargetFramework)' == 'netcoreapp3.1'" />
```

**Comment by devlead** [comment]
Now also applied the same fix as cake-build/bakery#153 issues found in cake-build/bakery#152

File: src/Cake/Infrastructure/Scripting/ReferenceAssemblyResolver.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+#endif
+                {
+                    Assembly assembly;
+                    try
+                    {
+                        assembly = Assembly.Load(System.IO.Path.GetFileNameWithoutExtension(reference.FilePath));
+                    }
+                    catch (Exception ex)
+                    {
+                        _log.Debug(log => log("Failed to load {0}\r\n{1}", reference.FilePath, ex));
```
Comment (gitfool) [inline src/Cake/Infrastructure/Scripting/ReferenceAssemblyResolver.cs:39]: Should `\r\n` be replaced with something portable? Actually, I see `\r\n` is sprinkled throughout the Cake source, perhaps inconsistently?

**Review by bjorkstromm — APPROVED** [review]
LGTM

File: src/Cake.Core/Scripting/ScriptConventions.cs [hunk 3]
```diff
@@ -20,18 +20,21 @@ public sealed class ScriptConventions : IScriptConventions
...
 
         /// <summary>
         /// Initializes a new instance of the <see cref="ScriptConventions"/> class.
         /// </summary>
         /// <param name="fileSystem">The file system.</param>
         /// <param name="loader">The assembly loader.</param>
         /// <param name="runtime">The Cake runtime.</param>
-        public ScriptConventions(IFileSystem fileSystem, IAssemblyLoader loader, ICakeRuntime runtime)
+        /// <param name="referenceAssemblyResolver">The reference assembly resolver.</param>
+        public ScriptConventions(IFileSystem fileSystem, IAssemblyLoader loader, ICakeRuntime runtime, IReferenceAssemblyResolver referenceAssemblyResolver)
```
Comment (gep13) [inline src/Cake.Core/Scripting/ScriptConventions.cs:32]: @devlead does this constitute a breaking change?

Comment (devlead) [reply src/Cake.Core/Scripting/ScriptConventions.cs:32]: Yes, and we've bumped the minimal Cake version already.

Though think very few use this part of Cake externally, as it's compiling part of scripts.

Comment (gep13) [reply src/Cake.Core/Scripting/ScriptConventions.cs:32]: Ok, makes sense. I will update the linked issue with the breaking change label, just so it is clear.

**Review by gep13 — APPROVED** [review]
LGTM!

File: src/Cake/Infrastructure/Scripting/ReferenceAssemblyResolver.cs (hunk 2 above)
Comment (gep13) [reply src/Cake/Infrastructure/Scripting/ReferenceAssemblyResolver.cs:39]: @gitfool This is definitely something that can be improved, but given that this is a debug log, I don't think this is a showstopper.  If you have found this being used in other places, would you be up for a maintenance PR to fix it up in all places?  For now, I will resolve this conversation.

**Comment by gep13** [comment]
@devlead your changes have been merged, thanks for your contribution 👍

**Pull request merged**